File: librtEdge/cpp/LVCAdmin.cpp
/******************************************************************************
*
*  LVCAdmin.cpp
*
*  REVISION HISTORY:
*     13 NOV 2014 jcs  Created (from Subscribe.cpp)
*     . . .
*     14 JUN 2022 jcs  Build 55: LVCAdmin.cpp
*     24 AUG 2023 jcs  Build 64: Named Schemas
*     26 JAN 2024 jcs  Build 68: -db / AddInterestList()
*
*  (c) 1994-2024, Gatea, Ltd.
******************************************************************************/
#include <librtEdge.h>

using namespace RTEDGE;

/////////////////////////////////////
// Version
/////////////////////////////////////
const char *LVCAdminID()
{
   static std::string s;
   const char   *sccsid;

   // Once

   if ( !s.length() ) {
      char bp[K], *cp;

      cp  = bp;
      cp += sprintf( cp, "@(#)LVCAdmin Build %s ", _MDD_LIB_BLD );
      cp += sprintf( cp, "%s %s Gatea Ltd.\n", __DATE__, __TIME__ );
      cp += sprintf( cp, rtEdge::Version() );
      s   = bp;
   }
   sccsid = s.data();
   return sccsid+4;
}

/////////////////////////////////////
//
//   c l a s s   M y A d m i n
//
/////////////////////////////////////
class MyAdmin : public LVCAdmin
{
public:
   size_t _nAck;
   bool   _bUP;

   ///////////////////////////////////
   // Constructor / Destructor
   ///////////////////////////////////
public:
   MyAdmin( const char *adm ) :
      LVCAdmin( adm ),
      _nAck( 0 ),
      _bUP( 0 )
   { ; }

   ///////////////////////////////////
   // RTEDGE::LVCAdmin Notifications
   ///////////////////////////////////
   virtual void OnConnect( const char *msg, bool bUP )
   {
      const char *ty = bUP ? "UP" : "DOWN";

      _bUP = bUP;
      ::fprintf( stdout, "CONN.%s : %s\n", ty, msg );
      ::fflush( stdout );
   }

   virtual bool OnAdminACK( bool bAdd, const char *svc, const char *tkr )
   {
      const char *ty = bAdd ? "ADD" : "DEL";

      ::fprintf( stdout, "ACK.%s : ( %s,%s )\n", ty, svc, tkr );
      ::fflush( stdout );
      _nAck += 1;
      return true;
   }

   virtual bool OnAdminNAK( bool bAdd, const char *svc, const char *tkr )
   {
      const char *ty = bAdd ? "ADD" : "DEL";

      ::fprintf( stdout, "NAK.%s : ( %s,%s )\n", ty, svc, tkr );
      ::fflush( stdout );
      _nAck += 1;
      return true;
   }

}; // class MyAdmin


//////////////////////////
// main()
//////////////////////////
int main( int argc, char **argv )
{
   Strings     tdb;
   std::string s;
   char        sTkr[K], *cp, *rp;
   bool        bAdd, aOK, bCfg;
   size_t      n, nt, na;
   FILE       *fp;
   const char *db, *cmd, *svr, *svc, *tkr, *schema;

   /////////////////////
   // Quickie checks
   /////////////////////
   if ( argc > 1 && !::strcmp( argv[1], "--version" ) ) {
      printf( "%s\n", LVCAdminID() );
      return 0;
   }
   db     = NULL;
   cmd    = "ADD";
   svr    = "localhost:8775";
   svc    = "bloomberg"; 
   tkr    = "";
   schema = "";
   bCfg   = ( argc < 2 ) || ( argc > 1 && !::strcmp( argv[1], "--config" ) );
   if ( bCfg ) {
      s  = "Usage: %s \\ \n";
      s += "       [ -db <LVC d/b for AddInterestList()> ] \\ \n";
      s += "       [ -c  <ADD | DEL> ] \\ \n";
      s += "       [ -h  <LVC Admin host:port> ] \\ \n";
      s += "       [ -s  <Service> ] \\ \n";
      s += "       [ -t  <Ticker : CSV or Filename flat ASCII> ] \\ \n";
      s += "       [ -x  <Schema Name> ] \\ \n";
      printf( s.data(), argv[0] );
      printf( "   Defaults:\n" );
      printf( "      -db      : <empty>\n" );
      printf( "      -c       : %s\n", cmd );
      printf( "      -h       : %s\n", svr );
      printf( "      -s       : %s\n", svc );
      printf( "      -t       : <empty>\n" );
      printf( "      -x       : <empty>\n" );
      return 0;
   }

   /////////////////////
   // cmd-line args
   /////////////////////
   for ( int i=1; i<argc; i++ ) {
      aOK = ( i+1 < argc );
      if ( !aOK )
         break; // for-i
      if ( !::strcmp( argv[i], "-db" ) )
         db = argv[++i];
      else if ( !::strcmp( argv[i], "-c" ) )
         cmd = argv[++i];
      else if ( !::strcmp( argv[i], "-h" ) )
         svr = argv[++i];
      else if ( !::strcmp( argv[i], "-s" ) )
         svc = argv[++i];
      else if ( !::strcmp( argv[i], "-t" ) )
         tkr = argv[++i];
      else if ( !::strcmp( argv[i], "-x" ) )
         schema = argv[++i];
   }
   bAdd = ( ::strcmp( cmd, "DEL" ) != 0 );

   ////////////////
   // Tickers 
   ////////////////
   if ( (fp=::fopen( tkr, "r" )) ) {
      while( ::fgets( (cp=sTkr), K, fp ) ) {
         ::strtok( sTkr, "#" );
         cp += ( strlen( sTkr ) - 1 );
         for ( ; cp > sTkr; ) {
            if ( ( *cp == '\r' ) || 
                 ( *cp == '\n' ) ||
                 ( *cp == '\t' ) || 
                 ( *cp == ' ' ) )
            {
               cp--;
               continue; // for-i
            }
            break; // for-cp
         }
         cp[1] = '\0';
         if ( strlen( sTkr ) )
            tdb.push_back( std::string( sTkr ) );
      }
      ::fclose( fp );
   }
   else {
      s   = tkr;
      tkr = ::strtok_r( (char *)s.data(), ",", &rp );
      for ( ; tkr && strlen( tkr ); tkr=::strtok_r( NULL, ",", &rp ) )
         tdb.push_back( std::string( tkr ) );
   }
   if ( !(nt=tdb.size()) ) {
      ::fprintf( stdout, "No ticker specified; Exitting ...\n" );
      ::fflush( stdout );
   }

   ////////////////
   // Run until ACK'ed
   ////////////////
   MyAdmin      adm( svr );
   const char **tkrs;
   size_t       sz;

   sz   = nt * sizeof( const char * );
   tkrs = (const char **)new char[sz];
   for ( n=0; n<nt; tkrs[n] = tdb[n].data(), n++ );
   tkrs[nt] = '\0';
   if ( bAdd ) {
      if ( db ) {
         LVC lvc( db );

         na = adm.AddFilteredTickers( lvc, svc, tkrs, schema );
         ::fprintf( stdout, "%ld of %ld Adding ...\n", na, nt );
         nt = na;
      }
      else
         adm.AddTickers( svc, tkrs, schema );
   }
   else
      adm.DelTickers( svc, tkrs );
   delete tkrs;
   /*
    * Wait up to 5 secs for all ACK's
    */
   for ( n=0; ( adm._nAck < nt ) && n<50; rtEdge::Sleep( 0.100 ), n++ );
   ::fprintf( stdout, "%ld of %ld ACK's received\n", adm._nAck, nt );
   adm.Stop();
   ::fprintf( stdout, "Done!!\n" );
   ::fflush( stdout );
   return 1;
}

File: librtEdge/py/examples/LVCAdmin.py
#!/usr/bin/python
#################################################################
#
#  LVCAdmin.py
#     LVCAdmin tester
#
#  REVISION HISTORY:
#     20 JUL 2022 jcs  Created
#     14 AUG 2023 jcs  Python 2 / 3
#      4 SEP 2023 jcs  Named Schema; args like LVCAdmin.cpp; DelTickers()
#      5 FEB 2025 jcs  Biuld 14: LVCAdmin.OnXxx()
#
#  (c) 1994-2025, Gatea Ltd.
#################################################################
import sys, time
import libMDDirect

_CDB = { True : 'UP',  False : 'DOWN' }
_ADB = { True : 'ADD', False : 'DEL' }

def Log( msg ):
   if msg:
      libMDDirect.Log( msg )
   return

class MyLVCAdmin( libMDDirect.LVCAdmin ):
   ################################
   # Constructor
   #
   # @param adm : host:port of LVC Admin Channel
   ################################
   def __init__( self, adm ):
      libMDDirect.LVCAdmin.__init__( self, adm )

   ########################
   # @override : Called asynchronously when we connect or disconnect
   #
   # @param msg - Textual description of connect event
   # @param bUP - True if UP; False if DOWN
   ########################
   def OnConnect( self, msg, bUP ):
      Log( 'CONN.{%s} : %s' % ( _CDB[bUP], msg ) )

   ########################
   # @override : Called asynchronously when an ACK message arrives
   #  
   # @param bAdd - true if ADD; false if DEL
   # @param svc - Service Name
   # @param tkr - Ticker Name 
   ########################
   def OnAdminACK( self, bAdd, svc, tkr ):
      Log( 'ACK.%s .{%s,%s}' % ( _ADB[bAdd], svc, tkr ) )

   ########################
   # @override : Called asynchronously when an NAK message arrives
   #  
   # @param bAdd - true if ADD; false if DEL
   # @param svc - Service Name
   # @param tkr - Ticker Name 
   ########################
   def OnAdminNAK( self, bAdd, svc, tkr ):
      Log( 'NAK.%s .{%s,%s}' % ( _ADB[bAdd], svc, tkr ) )


############################################
#
# main()
#
############################################
if __name__ == "__main__":
   #
   # -h <host:port> -c <cmd> -s <service> -t <ticker> -x <Schema>
   #
   Log( libMDDirect.Version() )
   argc   = len( sys.argv )
   cmd    = "ADD"
   adm    = "localhost:8775"
   svc    = "bloomberg"
   tkr    = ""
   schema = ""
   if argc < 2:
      s  = [ 'Usage: %s \\' % sys.argv[0],
             '       [ -c  <ADD | DEL | REFRESH> ] \\',
             '       [ -h  <LVC Admin host:port> ] \\',
             '       [ -s  <Service> ] \\',
             '       [ -t  <Ticker : CSV or Filename flat ASCII> ] \\',
             '       [ -x  <Schema Name> ]',
             '   Defaults:',
             '      -c       : %s' % cmd ,
             '      -h       : %s' % adm,
             '      -s       : %s' % svc,
             '      -t       : <empty>',
             '      -x       : <empty>',
           ]
      print( '\n'.join( s ) )
      sys.exit()
   for i in range( 1,argc,2 ):
      sel = sys.argv[i]
      try:    val = sys.argv[i+1]
      except: break ## for-i
      if sel == '-h':   adm    = val
      elif sel == '-s': svc    = val
      elif sel == '-t': tkr    = val
      elif sel == '-x': schema = val
      elif sel == '-c': cmd    = val
   if not adm.count( ':' ): adm += ':8775'
   try:
      fp   = open( tkr, 'rb' )
      tkrs = [ tt.strip('\n').strip('\r') for tt in fp.readlines() ]
      fp.close()
   except:
      tkrs = [ tt.strip('\n').strip('\r') for tt in tkr.split(',') ]
   #
   # Open LVC
   #
   #
   # Rock and Roll
   #
   lvc = MyLVCAdmin( adm )
   Log( 'LVCAdmin to %s' % adm )
   if cmd == 'REFRESH':
      lvc.RefreshTickers( svc, tkrs )
   elif cmd == 'DEL':
      lvc.DelTickers( svc, tkrs, schema )
   elif cmd == 'ADD':
      lvc.AddTickers( svc, tkrs, schema )
   else:
      Log( 'Unknown command %s' % cmd )
      cmd = None
   if cmd:
      Log( '%d tickers %s-ed' % ( len( tkrs ), cmd ) )
   py3 = libMDDirect.IsPY3()
   msg = 'Hit <ENTER> to terminate\n'
   if py3: notUsed = input( msg )
   else:   raw_input( msg )
   Log( lvc.Close() )
   Log( 'Done!!' )
